lib/chef/provisioning/aws_driver/aws_resource_with_entry.rb

```diff
@@ -2,21 +2,78 @@
 
 # Common AWS resource - contains metadata that all AWS resources will need
 class Chef::Provisioning::AWSDriver::AWSResourceWithEntry < Chef::Provisioning::AWSDriver::AWSResource
+  #
+  # Dissociate the ID of this object from Chef.
+  #
+  # @param action_handler [Chef::Provisioning::ActionHandler] The action handler,
+  #        which handles progress reporting, update reporting ("little green text")
+  #        and dry run.
+  #
   def delete_managed_entry(action_handler)
-    managed_entry_store.delete(self.class.resource_name, name, action_handler)
+    if should_have_managed_entry?
+      managed_entry_store.delete(self.class.resource_name, name, action_handler)
+    end
+  end
+
+  #
+  # Save the ID of this object to Chef.
+  #
+  # @param aws_object [AWS::EC2::Core] The AWS object containing the ID.
+  # @param action_handler [Chef::Provisioning::ActionHandler] The action handler,
+  #        which handles progress reporting, update reporting ("little green text")
+  #        and dry run.
+  # @param existing_entry [Chef::Provisioning::ManagedEntry] The existing entry
+  #        (if any).  If this is passed in, and no values are changed, we will
+  #        not attempt to update it (this prevents us from retrieving it twice).
+  #
+  def save_managed_entry(aws_object, action_handler, existing_entry: nil)
+    if should_have_managed_entry?
+      managed_entry = existing_entry ||
+                      managed_entry_store.new_entry(self.class.resource_name, name)
+      updated = update_managed_entry(aws_object, managed_entry)
+      if updated || !existing_entry
+        managed_entry.save(action_handler)
+      end
+    end
   end
 
-  def save_managed_entry(aws_object, action_handler)
-    managed_entry = managed_entry_store.new_entry(self.class.resource_name, name)
-    managed_entry.reference = { self.class.managed_entry_id_name => aws_object.public_send(self.class.aws_sdk_class_id) }
-    managed_entry.driver_url = driver.driver_url
-    managed_entry.save(action_handler)
+  def get_id_from_managed_entry
+    if should_have_managed_entry?
+      entry = managed_entry_store.get(self.class.managed_entry_type, name)
+      if entry
+        driver = self.driver
+        if entry.driver_url != driver.driver_url
+          # TODO some people don't send us run_context (like Drivers).  We might need
+          # to exit early here if the driver_url doesn't match the provided driver.
+          driver = run_context.chef_provisioning.driver_for(entry.driver_url)
+        end
+        [ driver, entry.reference[self.class.managed_entry_id_name], entry ]
+      end
+    end
   end
 
   protected
 
+  #
+  # Update an existing ManagedEntry object.
+  #
+  # @return true if the entry was changed, and false if not
+  #
+  def update_managed_entry(aws_object, managed_entry)
```

Comment:
I think we can achieve FULL unit test coverage of this method.  If we add new fields to the managed entry, we can add unit tests that show an old managed entry is correctly updated with the new fields.